feat(oxcaml): add the library_parameter stanza

maiste opened the pull request:
This PR is the first of a set of four PRs related to the introduction of Parameterized Libraries in Dune.

This feature is only available with OxCaml until it is upstream to the compiler. 

It contains multiples changes:
- Support for the `library_parameter` stanza
- Test of the newly introduced stanza. The tests will be updated with the next PRs.
- The documentation related to the new stanza.

Closes #12084

maiste committed:
feat: guard the library_parameter with oxcaml extension

It uses the (use oxcaml 1.0) extension to make sure the experimental
feature is not exposed directly to Dune users.

Signed-off-by: Etienne Marais <self@maiste.fr>

maiste committed:
doc: add library_parameter docs

Signed-off-by: Etienne Marais <self@maiste.fr>

maiste commented on src/dune_rules/stanzas/parameter.ml:
```diff
@@ -0,0 +1,138 @@
+          | Error user_message ->
+            User_error.raise
+              ~loc
+              [ Pp.textf "Invalid library name."
+              ; Pp.text
+                  "Public library names don't have this restriction. You can either \
+                   change this public name to be a valid library name or add a \"name\" \
+                   field with a valid library name."
```
I'll keep the `library_parameter` name to not create confusion later with the `parameters` field :+1:

maiste commented on src/dune_rules/utop.ml:
```diff
@@ -95,6 +96,7 @@ let add_stanza db ~dir (acc, pps) stanza =
          let info = Lib.info lib in
          match Lib_info.kind info with
          | Normal -> Appendable_list.cons lib acc, pps
+         | Parameter -> failwith "TODO arthur"
```
I'll add it as `Normal` and left a CR comment to remember to rework this when all the PRs would be merged.

maiste commented on doc/reference/dune/library_parameter.rst:
```diff
@@ -0,0 +1,112 @@
+    Specifies what modules are part of the library parameter. By default, Dune will use
+    all the ``.mli`` files in the same directory as the ``dune`` file. This
+    includes ones present in the file system as well as ones generated by user
+    rules. You can restrict this list by using a ``(modules <modules>)`` field.
```
Should we go in the direction where we keep the `modules` field or if, by default, `library_parameter` only take the `parameter-name` and exclude the rest of the modules? My point of view is that `modules` would keep Dune users in familiar waters but taking `parameter-name` would remove the risk to specify multiple modules... I'm not sure about the best behaviour to adapt here. 

In the rework version, I have simplified the phrasing for our need.

WDYT @rgrinberg?

rgrinberg commented:
@maiste is this ready for another round of review?

maiste commented:
@rgrinberg gentle ping. I have rebased the PR with `main` to make the CI happy.  I'll rework the commit history after review to make more readable commits.

rgrinberg reviewed (approved):
LGTM. I think the library_parameter stanza is overly permissive at the moment. Might be worth documenting this in the tests.

maiste committed:
feat: add library_parameter stanza

Co-authored-by: Rudi Grinberg <me@rgrinberg.com>
Co-authored-by: ArthurW <arthur@tarides.com>

Signed-off-by: Etienne Marais <self@maiste.fr>

maiste committed:
test(oxcaml): parameter tests

This commit adds a first batch of tests regarding the parametrized
library. These tests are extended in future commits to add more checks
as it requires extra behaviour to check it deeper.

Signed-off-by: Etienne Marais <self@maiste.fr>

rgrinberg commented on test/blackbox-tests/test-cases/oxcaml/library_parameter.t:
```diff
@@ -0,0 +1,143 @@
+  > (library_parameter
+  >  (public_name param.intf)
+  >  (name param_intf)
+  >  (modules param_intf))
```
Can you add a test for a library parameter with more than one module?

rgrinberg commented:
Sorry my previous comment was a bit opaque. I expanded on the fields I think are unnecessary and which tests I would like to see.

maiste committed:
feat: export some library buildable decoders

Signed-off-by: Etienne Marais <self@maiste.fr>

maiste committed:
feat: use exported buildable fields internally

Signed-off-by: Etienne Marais <self@maiste.fr>

maiste commented on src/dune_rules/stanzas/parameter.ml:
```diff
@@ -0,0 +1,137 @@
+     and+ package = field_o "package" (located Stanza_common.Pkg.decode)
+     and+ enabled_if = Enabled_if.decode ~allowed_vars:Any ~since:None ()
+     and+ synopsis = field_o "synopsis" string
+     and+ stdlib =
```
Even if you want to use `Base` as the standard library? Otherwise, I can remove it.

Alizter commented on test/blackbox-tests/test-cases/oxcaml/library_parameter.t:
```diff
@@ -61,9 +61,35 @@ We rebuild it and ensure we have the new parameter available.
+  > type t = string
+  > EOF
+
+  $ dune build
```
I'm a bit confused, it doesn't seem like this failed?

maiste replied:
Yep, and that's a problem I need to fix for this PR.

maiste replied:
I have pushed the fix: https://github.com/ocaml/dune/pull/11963/files#diff-97a3106adc9720a309352f46b8d89a0bdf768edd1ebf5709d49f925fce3aa07fR438

maiste committed:
chore: add a CHANGES.md entry

Signed-off-by: Etienne Marais <self@maiste.fr>
